README.md
## Creatures Revived

Copyright (c) 2015-2016 BlockMen <blockmen2015@gmail.com>

Version: 1.1

---
### Description:

A revived version of BlockMen's Creatures MOB-Engine mod.

This mod provides an engine, that handles the base function for MOB in Minetest.
It offers an easy way to register MOB and allows to custom handling for the needs
of each mob. This engine aims to be a solid base, that has a good balance between
performance and functionality.
See API.txt for more informations on how to use this engine for mobs.

![screenshot](screenshot.png)

---
### Licensing:

- Code: [modified Zlib](LICENSE.txt) : (c) Copyright 2015-2016 BlockMen
- Sounds:
	- cmer_hit_01.ogg: CC0
- All other media (textures and sounds): (c) Copyright (2014-2016) BlockMen; CC-BY-SA 3.0

---
### Links:

- [Forum](https://forum.minetest.net/viewtopic.php?t=26684)
- [Git repo](https://github.com/AntumMT/mod-cmer)
- [API](https://antummt.github.io/mod-cmer/docs/api.html)
- [Changelog](CHANGES.txt)
- [TODO](TODO.txt)
- Original Creatures MOB-Engine by BlockMen:
	- [Forum](https://forum.minetest.net/viewtopic.php?t=8638)
	- [Git repo](https://github.com/BlockMen/cme)
- Other derivatives:
	- [Mob Engine](https://github.com/minetest-mods/mob-engine)


CHANGES.txt

revived-1.1
- Renamed to "cmer" to discern from other versions.
- Removed built-in items.
- Removed built-in spawnegg support.
- entities whose lifetime expires do not drop items
- default sound plays when hit (can be disabled in definition with sounds.play_default_hit = false)
- entity's lifetime doesn't expire in modes "follow" & "panic"
- entity's life timer is reset when attacked
- can set custom on_death method
- 'has_kockback' deprecated, use 'has_knockback'
- does not automatically prepend colon (:) when registering a mob name
- option to set nametag
- griefing is disabled unless 'mobs_griefing' setting enabled
- owned entities can be picked up with mobs:lasso
- owned entities cannot be killed by other players


revived-1.0
- Forked from original Creatures MOB-Engine code by BlockMen.
- Updated for Minetest 5.x API.
- Removed built-in modpack mobs (engine only).
- Some nil check fixes.
- Added default combat speed.
- Added "owner" attribute.
- Fixed so owned/tamed mobs don't die.


init.lua
--= Zombie for Creatures MOB-Engine (cme) =--
-- Copyright (c) 2015-2016 BlockMen <blockmen2015@gmail.com>
--
-- init.lua
--
-- This software is provided 'as-is', without any express or implied warranty. In no
-- event will the authors be held liable for any damages arising from the use of
-- this software.
--
-- Permission is granted to anyone to use this software for any purpose, including
-- commercial applications, and to alter it and redistribute it freely, subject to the
-- following restrictions:
--
-- 1. The origin of this software must not be misrepresented; you must not
-- claim that you wrote the original software. If you use this software in a
-- product, an acknowledgment in the product documentation is required.
-- 2. Altered source versions must be plainly marked as such, and must not
-- be misrepresented as being the original software.
-- 3. This notice may not be removed or altered from any source distribution.
--


cmer_zombie = {}
modname = core.get_current_modname()
modpath = core.get_modpath(modname)

local scripts = {
	"settings",
	"item",
	"entity",
}

for _, script in ipairs(scripts) do
	dofile(modpath .. "/" .. script .. ".lua")
end


common.lua
--= Creatures MOB-Engine (cme) =--
-- Copyright (c) 2015-2016 BlockMen <blockmen2015@gmail.com>
--
-- common.lua
--
-- This software is provided 'as-is', without any express or implied warranty. In no
-- event will the authors be held liable for any damages arising from the use of
-- this software.
--
-- Permission is granted to anyone to use this software for any purpose, including
-- commercial applications, and to alter it and redistribute it freely, subject to the
-- following restrictions:
--
-- 1. The origin of this software must not be misrepresented; you must not
-- claim that you wrote the original software. If you use this software in a
-- product, an acknowledgment in the product documentation is required.
-- 2. Altered source versions must be plainly marked as such, and must not
-- be misrepresented as being the original software.
-- 3. This notice may not be removed or altered from any source distribution.
--


-- constants
nullVec = {x = 0, y = 0, z = 0}
DEGTORAD = math.pi / 180.0

-- common functions
function cmer.sumChances(tab)
	local psum = 0
	for s,w in pairs(tab) do
		psum = psum + ((tonumber(w) or w.chance or 0))
	end
	return psum
end

function cmer.rnd(tab, errval)
	if not errval then
		errval = false
	end

	local res = 1000000000
	local rn = math.random(0, res - 1)
	local retval = nil

	local psum = 0
	for s,w in pairs(tab) do
		psum = psum + ((tonumber(w) or w.chance or 0) * res)
		if psum > rn then
			retval = s
			break
		end
	end

	return retval
end

function throw_error(msg)
	core.log("error", "#Creatures: ERROR: " .. msg)
end

function throw_warning(msg)
	core.log("warning", "#Creatures: WARNING: " .. msg)
end

function cmer.compare_pos(pos1, pos2)
	if not pos1 or not pos2 then
		return
	end
	if pos1.x == pos2.x and pos1.y == pos2.y and pos1.z == pos2.z then
		return false
	end
	return true
end

function cmer.findTarget(search_obj, pos, radius, search_type, ignore_mob, xray, no_count)
	local player_near = false
	local mobs = {}
	for  _,obj in ipairs(core.get_objects_inside_radius(pos, radius)) do
		if obj ~= search_obj then
			if xray or core.line_of_sight(pos, obj:get_pos()) == true then
				local is_player = obj:is_player()
				if is_player then
					player_near = true
					if no_count == true then
						return {}, true
					end
				end
				local entity = obj:get_luaentity()
				local isItem = (entity and entity.name == "__builtin:item") or false
				local ignore = (entity and entity.mob_name == ignore_mob and search_type ~= "mates") or false

				if search_type == "all" then
					if not isItem and not ignore then
						table.insert(mobs, obj)
					end
				elseif search_type == "hostile" then
					if not ignore and (entity and entity.hostile == true) or is_player then
						table.insert(mobs, obj)
					end
				elseif search_type == "nonhostile" then
					if entity and not entity.hostile and not isItem and not ignore then
						table.insert(mobs, obj)
					end
				elseif search_type == "player" then
					if is_player then
						table.insert(mobs, obj)
					end
				elseif search_type == "mate" then
					if not isItem and (entity and entity.mob_name == ignore_mob) then
						table.insert(mobs, obj)
					end
				end
			end
		end --for
	end

	return mobs,player_near
end

function cmer.dropItems(pos, drops)
	if not pos or not drops then
		return
	end

	-- convert drops table
	local tab = {}
	for _,elem in pairs(drops) do
		local name = tostring(elem[1])
		local v = elem[2]
		local chance = elem.chance
		local amount = ""
		-- check if drops depending on defined chance
		if name and chance then
			local ct = {}
			ct[name] = chance
			ct["_fake"] = 1 - chance
			local res = cmer.rnd(ct)
			if res == "_fake" then
				name = nil
			end
		end
		-- get amount
		if name and v then
			if type(v) == "table" then
				amount = math.random(v.min or 1, v.max or 1) or 1
			elseif type(v) == "number" then
				amount = v
			end
			if amount > 0 then
				amount = " " .. amount
			end
		end
		if name then
			local obj = core.add_item(pos, name .. amount)
			if not obj then
				throw_error("Could not drop item '" .. name .. amount .. "'")
			end
		end
	end
end


entity.lua

-- base specs
local zombie = {
	modes = {
		idle = {chance=0.7, duration=3, update_yaw=6},
		walk = {chance=0.3, duration=5.5, moving_speed=1.5},
		attack = {chance=0, moving_speed=2.5},
	},
	mesh = "creatures_zombie.b3d",
	collisionbox = {-0.25, -0.01, -0.25, 0.25, 1.65, 0.25},
	rotation = -90.0,
	animations = {
		idle = {start=0, stop=80, speed=15},
		walk = {start=102, stop=122, speed=15.5},
		attack = {start=102, stop=122, speed=25},
		death = {start=81, stop=101, speed=28, loop=false, duration=2.12},
	},
	sounds = {
		on_damage = {name="creatures_zombie_hit", gain=0.4, distance=10},
		on_death = {name="creatures_zombie_death", gain=0.7, distance=10},
		swim = {name="creatures_splash", gain=1.0, distance=10},
		random = {
			idle = {name="creatures_zombie", gain=0.7, distance=12},
		},
	},
	drops = {
		{"creatures:rotten_flesh", {min=1, max=2}, chance=0.7},
	},
	nodes = {
		spawn_on = {"default:stone", "default:dirt_with_grass", "default:dirt",
			"default:cobblestone", "default:mossycobble", "group:sand"},
	},
}


creatures.register_mob({
	-- general
	name = ":creatures:zombie",
	nametag = creatures.feature_nametags and "Zombie" or nil,
	stats = {
		hp = 20,
		lifetime = cmer_zombie.lifetime,
		can_jump = 1,
		can_swim = true,
		can_burn = true,
		has_falldamage = true,
		has_knockback = true,
		light = {min=0, max=8},
		hostile = true,
	},

	modes = zombie.modes,

	model = {
		mesh = zombie.mesh,
		textures = {"creatures_zombie.png"},
		collisionbox = zombie.collisionbox,
		rotation = zombie.rotation,
		animations = zombie.animations,
	},

	sounds = zombie.sounds,

	combat = {
		attack_damage = 1,
		attack_speed = 0.6,
		attack_radius = 1.1,

		search_enemy = true,
		search_timer = 2,
		search_radius = 12,
		search_type = "player",
	},

	spawning = {
		abm_nodes = zombie.nodes,
		abm_interval = cmer_zombie.spawn_interval,
		abm_chance = cmer_zombie.spawn_chance,
		max_number = 1,
		number = 2,
		light = {min=0, max=8},
		height_limit = {min=-200, max=50},

		spawner = {
			description = "Zombie Spawner",
			range = 8,
			number = 6,
			light = {min=0, max=8},
		}
	},

	drops = zombie.drops,
})

if core.global_exists("asm") then
	asm.addEgg({
		name = "zombie",
		spawn = "creatures:zombie",
		inventory_image = "creatures_zombie_inv.png",
		ingredient = "creatures:rotten_flesh",
	})
end


-- harder zombie
creatures.register_mob({
	name = ":creatures:zombie_manic",
	nametag = creatures.feature_nametags and "Manic Zombie" or nil,

	stats = {
		hp = 75,
		lifetime = cmer_zombie.lifetime,
		can_jump = 1,
		can_swim = true,
		can_burn = true,
		has_falldamage = true,
		has_knockback = true,
		light = {min=0, max=8},
		hostile = true,
	},

	modes = zombie.modes,

	model = {
		mesh = zombie.mesh,
		textures = {"creatures_zombie_manic.png"},
		collisionbox = zombie.collisionbox,
		rotation = zombie.rotation,
		animations = zombie.animations,
	},

	sounds = zombie.sounds,

	combat = {
		attack_damage = 15,
		attack_speed = 0.6,
		attack_radius = 1.1,

		search_enemy = true,
		search_timer = 2,
		search_radius = 20,
		search_type = "player",
	},
	drops = zombie.drops,

	spawning = {
		abm_nodes = zombie.nodes,
		abm_interval = cmer_zombie.spawn_interval + math.floor(cmer_zombie.spawn_interval / 2),
		abm_chance = cmer_zombie.spawn_chance + math.floor(cmer_zombie.spawn_chance / 2),
		max_number = 1,
		number = 1,
		light = {min=0, max=8},
		height_limit = {min=-32000, max=500},
	},
})


settings.lua
--
--

--- Creatures Revived Settings
--
--  @module settings.lua


cmer.debug = core.settings:get_bool("enable_debug_mods", false)


--- Disables attacking players.
--
--  @setting only_peaceful_mobs
--  @settype bool
--  @default false
cmer.allow_hostile = core.settings:get_bool("only_peaceful_mobs") ~= true

--- Enables damage.
--
--  Global setting.
--
--  @setting enable_damage
--  @settype bool
--  @default false
cmer.enable_damage = core.settings:get_bool("enable_damage", false)

--- Enables creative mode.
--
--  Global setting.
--
--  @setting creative_mode
--  @settype bool
--  @default false
cmer.creative = core.settings:get_bool("creative_mode", false)


-- mobs specific settings

--- Enables particles when swimming.
--
--  @setting mobs.particles
--  @settype bool
--  @default false
cmer.particles = core.settings:get_bool("mobs.particles", false)

--- Displays nametags above mobs.
--
--  @setting mobs.nametags
--  @settype bool
--  @default false
cmer.nametags = core.settings:get_bool("mobs.nametags", false)

--- Determines if owned entities can be killed by non-owners.
--
--  @setting mobs.grief_owned
--  @settype bool
--  @default false
cmer.grief_owned = core.settings:get_bool("mobs.grief_owned", false)


functions.lua
--= Creatures MOB-Engine (cme) =--
-- Copyright (c) 2015-2016 BlockMen <blockmen2015@gmail.com>
--
-- functions.lua
--
-- This software is provided 'as-is', without any express or implied warranty. In no
-- event will the authors be held liable for any damages arising from the use of
-- this software.
--
-- Permission is granted to anyone to use this software for any purpose, including
-- commercial applications, and to alter it and redistribute it freely, subject to the
-- following restrictions:
--
-- 1. The origin of this software must not be misrepresented; you must not
-- claim that you wrote the original software. If you use this software in a
-- product, an acknowledgment in the product documentation is required.
-- 2. Altered source versions must be plainly marked as such, and must not
-- be misrepresented as being the original software.
-- 3. This notice may not be removed or altered from any source distribution.
--


-- Localizations
local rnd = math.random


local function knockback(selfOrObject, dir, old_dir, strengh)
	local object = selfOrObject
	if selfOrObject.mob_name then
		object = selfOrObject.object
	end
	local current_fmd = object:get_properties().automatic_face_movement_dir or 0
	object:set_properties({automatic_face_movement_dir=false})
	object:set_velocity(vector.add(old_dir, {x=dir.x * strengh, y=3.5, z=dir.z * strengh}))
	old_dir.y = 0
	core.after(0.4, function()
		object:set_properties({automatic_face_movement_dir=current_fmd})
		object:set_velocity(old_dir)
		selfOrObject.falltimer = nil
		if selfOrObject.stunned == true then
			selfOrObject.stunned = false
			if selfOrObject.can_panic == true then
				selfOrObject.target = nil
				selfOrObject.mode = "_run"
				selfOrObject.modetimer = 0
			end
		end
	end)
end

local function on_hit(me)
	core.after(0.1, function()
		me:set_texture_mod("^[colorize:#c4000099")
	end)
	core.after(0.5, function()
		me:set_texture_mod("")
	end)
end

local hasMoved = cmer.compare_pos

local function getDir(pos1, pos2)
	local retval
	if pos1 and pos2 then
		retval = {x=pos2.x - pos1.x, y=pos2.y - pos1.y, z=pos2.z - pos1.z}
	end
	return retval
end

local function getDistance(vec, fly_offset)
	if not vec then
		return -1
	end
	if fly_offset then
		vec.y = vec.y + fly_offset
	end
	return math.sqrt((vec.x)^2 + (vec.y)^2 + (vec.z)^2)
end

local findTarget = cmer.findTarget

local function update_animation(obj_ref, mode, anim_def)
	if anim_def and obj_ref then
		obj_ref:set_animation({x=anim_def.start, y=anim_def.stop}, anim_def.speed, 0, anim_def.loop)
	end
end

local function update_velocity(obj_ref, dir, speed, add)
	local velo = obj_ref:get_velocity()
	if not dir.y then
		dir.y = velo.y / speed
	end
	local new_velo = {x=dir.x * speed, y=dir.y * speed or velo.y ,z=dir.z * speed}
	if add then
		new_velo = vector.add(velo, new_velo)
	end
	obj_ref:set_velocity(new_velo)
end

local function getYaw(dirOrYaw)
	local yaw = 360 * rnd()
	if dirOrYaw and type(dirOrYaw) == "table" then
		yaw = math.atan(dirOrYaw.z / dirOrYaw.x) + math.pi^2 - 2
		if dirOrYaw.x > 0 then
			yaw = yaw + math.pi
		end
	elseif dirOrYaw and type(dirOrYaw) == "number" then
		-- here could be a value based on given yaw
	end

	return yaw
end

local dropItems = cmer.dropItems

local function killMob(me, def)
	if not def then
		if me then
			me:remove()
		end
	end
	local pos = me:get_pos()
	me:set_velocity(nullVec)
	me:set_properties({collisionbox=nullVec})
	me:set_hp(0)
	-- Must be non-zero otherwise mob is removed before death animation.
	me:set_hp(1)

	if def.sounds and def.sounds.on_death then
		local death_snd = def.sounds.on_death
		core.sound_play(death_snd.name, {
			pos = pos,
			max_hear_distance = death_snd.distance or 5,
			gain = death_snd.gain or 1,
			})
	end

	if def.model.animations.death then
		local dur = def.model.animations.death.duration or 0.5
		update_animation(me, "death", def.model.animations["death"])
		core.after(dur, function()
			me:remove()
		end)
	else
		me:remove()
	end
	if def.drops then
		if type(def.drops) == "function" then
			def.drops(me:get_luaentity())
		else
			dropItems(pos, def.drops)
		end
	end
end

local function limit(value, min, max)
	if value < min then
		return min
	end
	if value > max then
		return max
	end
	return value
end

local function calcPunchDamage(obj, actual_interval, tool_caps)
	local damage = 0
	if not tool_caps or not actual_interval then
		return 0
	end
	local my_armor = obj:get_armor_groups() or {}
	for group, _ in pairs(tool_caps.damage_groups) do
		damage = damage + (tool_caps.damage_groups[group] or 0) * limit(actual_interval / tool_caps.full_punch_interval, 0.0, 1.0) * ((my_armor[group] or 0) / 100.0)
	end
	return damage or 0
end

local function onDamage(self, hp, hitsound)
	local me = self.object
	local def = core.registered_entities[self.mob_name]
	hp = hp or me:get_hp()

	if hp <= 0 then
		self.stunned = true
			self:on_death()
			killMob(me, def)
	else
		on_hit(me) -- red flashing
		if def.sounds then
			if def.sounds.on_damage then
				local dmg_snd = def.sounds.on_damage
				core.sound_play(dmg_snd.name, {
					pos = me:get_pos(),
					max_hear_distance = dmg_snd.distance or 5,
					gain = dmg_snd.gain or 1,
					})
			end

			if hitsound and def.sounds.play_default_hit ~= false then
				core.sound_play(hitsound, {object=me})
			end
		end
	end
end

local function changeHP(self, value, hitsound)
	local me = self.object
	local hp = me:get_hp()
	hp = hp + math.floor(value)
	me:set_hp(hp)
	if value < 0 then
		onDamage(self, hp, hitsound)
	end
end

local function checkWielded(wielded, itemList)
	if not itemList then return false end

	for s, w in pairs(itemList) do
		if w == wielded then
			return true
		end
	end

	return false
end

local tool_uses = {0, 30, 110, 150, 280, 300, 500, 1000}
local function addWearout(player, tool_def)
	if not cmer.creative then
		local item = player:get_wielded_item()
		if tool_def and tool_def.damage_groups and tool_def.damage_groups.fleshy then
			local uses = tool_uses[tool_def.damage_groups.fleshy] or 0
			if uses > 0 then
				local wear = 65535 / uses
				item:add_wear(wear)
				player:set_wielded_item(item)
			end
		end
	end
end

local function spawnParticles(...)
end

if cmer.enable_particles then
	spawnParticles = function(pos, velocity, texture_str)
		local vel = vector.multiply(velocity, 0.5)
		vel.y = 0
		core.add_particlespawner({
			amount = 8,
			time = 1,
			minpos = vector.add(pos, -0.7),
			maxpos = vector.add(pos, 0.7),
			minvel = vector.add(vel, {x=-0.1, y=-0.01, z=-0.1}),
			maxvel = vector.add(vel, {x=0.1, y=0, z=0.1}),
			minacc = vector.new(),
			maxacc = vector.new(),
			minexptime = 0.8,
			maxexptime = 1,
			minsize = 1,
			maxsize = 2.5,
			texture = texture_str,
		})
	end
end



-- --
-- Default entity functions
-- --

cmer.on_punch = function(self, puncher, tflp, tc, dir)
	if self.stunned == true then
		return
	end

	local me = self.object
	local mypos = me:get_pos()

	changeHP(self, calcPunchDamage(me, tflp, tc) * -1, "cmer_hit_01")
	-- reset lifetimer when attacked
	self.lifetimer = 0
	if puncher then
		if self.hostile then
			self.mode = "attack"
			self.target = puncher
		end
		if tflp >= 0.45 and self.stunned == false then
			if self.has_knockback == true then
				local v = me:get_velocity()
				v.y = 0
				if not self.can_fly then
					me:set_acceleration({x=0, y=-15, z=0})
				end
				knockback(self, dir, v, 5)
				self.stunned = true
			end

			-- add wearout to weapons/tools
			addWearout(puncher, tc)
		end
	end
end

cmer.on_death = function(self, killer)
	if self.object:get_nametag_attributes().text then
		-- clear nametag
		self.object:set_nametag_attributes({text=nil})
	end
end

cmer.on_rightclick = function(self, clicker)
end

cmer.on_step = function(self, dtime)
	-- first get the relevant specs; exit if we don't know anything (1-3ms)
	local def = core.registered_entities[self.mob_name]
	if not def then
		throw_error("Can't load creature-definition")
		return
	end

	-- set default speed if not specified
	if def.combat then
		def.combat.attack_speed = def.combat.attack_speed or 1.0
	end

	-- timer updates
	if not self.owner then
		self.lifetimer = self.lifetimer + dtime
	end
	self.modetimer = self.modetimer + dtime
	self.soundtimer = self.soundtimer + dtime
	self.yawtimer = self.yawtimer + dtime
	self.nodetimer = self.nodetimer + dtime
	self.followtimer = self.followtimer + dtime
	if self.envtimer then
		self.envtimer = self.envtimer + dtime
	end
	if self.falltimer then
		self.falltimer = self.falltimer + dtime
	end
	if self.searchtimer then
		self.searchtimer = self.searchtimer + dtime
	end
	if self.attacktimer then
		self.attacktimer = self.attacktimer + dtime
	end
	if self.swimtimer then
		self.swimtimer = self.swimtimer + dtime
	end

	-- main
	if self.stunned == true then
		return
	end

	local no_death_modes = (self.mode == "attack" or self.mode == "follow" or self.mode == "panic")
	if self.lifetimer > def.stats.lifetime and not (no_death_modes and self.target) then
		-- reset life timer if engaging in combat
		self.lifetimer = 0
		--if not self.tamed or (self.tamed and def.stats.dies_when_tamed) then
		if not self.owner or def.stats.dies_when_tamed then
			self.object:remove()
			return
		end
	end


	-- localize some things
	local modes = def.modes
	local current_mode = self.mode
	local me = self.object
	local moved = false
	local current_pos = me:get_pos()
	if current_pos == nil then return end

	current_pos.y = current_pos.y + 0.5
	moved = hasMoved(current_pos, self.last_pos) or false

	local fallen = false

	-- Update pos and current node if necessary
	if moved == true or not self.last_pos then
		-- for falldamage
		if self.has_falldamage and self.last_pos and not self.in_water then
			local dist = math.abs(current_pos.y - self.last_pos.y)
			if dist > 0 then
				self.fall_dist = self.fall_dist - dist
				if not self.falltimer then
					self.falltimer = 0
				end
			end
		end

		self.last_pos = current_pos
		if self.nodetimer > 0.2 then
			self.nodetimer = 0
			local current_node = core.get_node_or_nil(current_pos)
			self.last_node = current_node
			if def.stats.light then
				local wtime = core.get_timeofday()
				local llvl = core.get_node_light({x=current_pos.x, y=current_pos.y + 0.5, z=current_pos.z}) or 0
				self.last_llvl = llvl
			end
		end
	else
		if (modes[current_mode].moving_speed or 0) > 0 then
			update_velocity(me, nullVec, 0)
			if modes["idle"] and not (current_mode == "attack" or current_mode == "follow") then
				current_mode = "idle"
				self.modetimer = 0
			end
		end
		if self.fall_dist < 0 then
			fallen = true
		end
	end

	if fallen then
		local falltime = tonumber(self.falltimer) or 0
		local dist = math.abs(self.fall_dist) or 0
		self.falltimer = 0
		self.fall_dist = 0
		fallen = false

		local damage = 0
		if dist > 3 and not self.in_water and falltime / dist < 0.2 then
			damage = dist - 3
		end

		-- damage by calced value
		if damage > 0 then
			changeHP(self, damage * -1)
		end
	end


	-- special mode handling
	-- check distance to target
	if self.target and self.followtimer > 0.6 then
		self.followtimer = 0
		local p2 = self.target:get_pos()
		local dir = getDir(current_pos, p2)
		local offset
		if self.can_fly then
			offset = modes["fly"].target_offset
		end
		local dist = getDistance(dir, offset)
		local radius
		if self.hostile and def.combat then
			radius = def.combat.search_radius
		elseif modes["follow"] then
			radius = modes["follow"].radius
		end
		if dist == -1 or dist > (radius or 5) then
			self.target = nil
			current_mode = ""
		elseif dist > -1 and self.hostile and dist < def.combat.attack_radius then
			-- attack
			if self.attacktimer > def.combat.attack_speed then
				self.attacktimer = 0
				if core.line_of_sight(current_pos, p2) == true then
					self.target:punch(me, 1.0, {
						full_punch_interval = def.combat.attack_speed,
						damage_groups = {fleshy=def.combat.attack_damage},
					})
	      end
				update_velocity(me, self.dir, 0)
			end
		else
			if current_mode == "attack" or current_mode == "follow" then
				self.dir = vector.normalize(dir)
				me:set_yaw(getYaw(dir))
				if self.in_water then
					self.dir.y = me:get_velocity().y
				end
				update_velocity(me, self.dir, modes[current_mode].moving_speed or 0)
			end
		end
	end

	-- search a target (1-2ms)
	if not self.target and ((self.hostile and def.combat.search_enemy) or modes["follow"]) and current_mode ~= "_run" then
		local timer
		if self.hostile then
			timer = def.combat.search_timer or 2
		elseif modes["follow"] then
			timer = modes["follow"].timer
		end
		if self.searchtimer > (timer or 4) then
			self.searchtimer = 0
			local targets = {}
			if self.hostile then
				targets = findTarget(me, current_pos, def.combat.search_radius, def.combat.search_type, def.combat.search_xray)
			else
				targets = findTarget(me, current_pos, modes["follow"].radius or 5, "player")
			end
			if #targets > 1 then
				self.target = targets[rnd(1, #targets)]
			elseif #targets == 1 then
				self.target = targets[1]
			end
			if self.target then
				if self.hostile and modes["attack"] then
					current_mode = "attack"
				else
					local name = self.target:get_wielded_item():get_name()
					if name and checkWielded(name, modes["follow"].items) == true then
						current_mode = "follow"
						self.modetimer = 0
					else
						self.target = nil
					end
				end
			end
		end
	end

	if current_mode == "eat" and not self.eat_node then
		local nodes = modes[current_mode].nodes
		local p = {x=current_pos.x, y=current_pos.y - 1, z=current_pos.z}
		local sn = core.get_node_or_nil(p)
		local eat_node
		for _, name in pairs(nodes) do
			if self.last_node ~= nil and name == self.last_node.name then
				eat_node = current_pos
				break
			elseif sn and sn.name == name then
				eat_node = p
				break
			end
		end

		if not eat_node then
			current_mode = "idle"
		else
			self.eat_node = eat_node
		end
	end


	-- further mode handling
	-- update mode
	if current_mode ~= "attack" and
			(current_mode == "" or self.modetimer > (modes[current_mode].duration or 4)) then
		self.modetimer = 0

		local new_mode = cmer.rnd(modes) or "idle"
		if new_mode == "eat" and self.in_water == true then
			new_mode = "idle"
		end
		if current_mode == "follow" and rnd(1, 10) < 3 then
			new_mode = current_mode
		elseif current_mode == "follow" then
			-- "lock" searching a little bit
			self.searchtimer = rnd(5, 8) * -1
			self.target = nil
		end
		current_mode = new_mode

		-- change eaten node when mode changes
		if self.eat_node then
			local n = core.get_node_or_nil(self.eat_node)
			local nnn = n.name
			local def = core.registered_nodes[n.name]
			local sounds
			if def then
				 if def.drop and type(def.drop) == "string" then
					 nnn = def.drop
				 elseif not def.walkable then
					 nnn = "air"
				 end
			end
			if nnn and nnn ~= n.name and core.registered_nodes[nnn] then
				core.set_node(self.eat_node, {name=nnn})
				if not sounds then
					sounds = def.sounds
				end
				if sounds and sounds.dug then
					core.sound_play(sounds.dug, {
						pos = self.eat_node,
						max_hear_distance = 5,
						gain = 1,
						})
				end
			end
			self.eat_node = nil
		end
	end

	-- mode has changes, do things
	if current_mode ~= self.last_mode then
		self.last_mode = current_mode

		local moving_speed = modes[current_mode].moving_speed or 0
		if moving_speed > 0 then
			local yaw = (getYaw(me:get_yaw()) + 90.0) * DEGTORAD
			me:set_yaw(yaw + 4.73)
			self.dir = {x=math.cos(yaw), y=0, z=math.sin(yaw)}
			if self.can_fly then
				if current_pos.y >= (modes["fly"].max_height or 50) and not self.target then
					self.dir.y = -0.5
				else
					self.dir.y = (rnd() - 0.5)
				end
			end

			-- reduce speed in water
			if self.in_water == true then
				moving_speed = moving_speed * 0.7
			end
		else
				self.dir = nullVec
		end

		update_velocity(me, self.dir, moving_speed)
		local anim_def = def.model.animations[current_mode]
		if self.in_water and def.model.animations["swim"] then
			anim_def = def.model.animations["swim"]
		end
		update_animation(me, current_mode, anim_def)
	end

	-- update yaw
	if current_mode ~= "attack" and current_mode ~= "follow" and
			(modes[current_mode].update_yaw or 0) > 0 and
			self.yawtimer > (modes[current_mode].update_yaw or 4) then
		self.yawtimer = 0
		local mod = nil
		if current_mode == "_run" then
			mod = me:get_yaw()
		end
		local yaw = (getYaw(mod) + 90.0) * DEGTORAD
		me:set_yaw(yaw + 4.73)
		local moving_speed = modes[current_mode].moving_speed or 0
		if moving_speed > 0 then
			self.dir = {x=math.cos(yaw), y=nil, z=math.sin(yaw)}
			update_velocity(me, self.dir, moving_speed)
		end
	end

	--swim
	if self.can_swim and self.swimtimer > 0.8 and self.last_node then
		self.swimtimer = 0
		local name = self.last_node.name
		if name then
			if name == "default:water_source" then
				self.air_cnt = 0
				local vel = me:get_velocity()
				update_velocity(me, {x=vel.x, y=0.9, z=vel.z}, 1)
				me:set_acceleration({x=0, y=-1.2, z=0})
				self.in_water = true
				-- play swimming sounds
				if def.sounds and def.sounds.swim then
					local swim_snd = def.sounds.swim
					core.sound_play(swim_snd.name, {
						pos = current_pos,
						gain = swim_snd.gain or 1,
						max_hear_distance = swim_snd.distance or 10,
						})
				end
				spawnParticles(current_pos, vel, "bubble.png")
			else
				self.air_cnt = self.air_cnt + 1
				if self.in_water == true and self.air_cnt > 5 then
					self.in_water = false
					if not self.can_fly then
						me:set_acceleration({x=0, y=-15, z=0})
					end
				end
			end
		end
	end

	-- Add damage when drowning or in lava
	if self.env_damage and self.envtimer > 1 and self.last_node then
		self.envtimer = 0
		local name = self.last_node.name
		if not self.can_swim and name == "default:water_source" then
			changeHP(self, -1)
		elseif self.can_burn then
			if name == "fire:basic_flame" or name == "default:lava_source" then
				changeHP(self, -4)
			end
		end

		-- add damage when light is too bright or too dark
		local tod = core.get_timeofday() * 24000
		if self.last_llvl and self.can_burn and self.last_llvl > (def.stats.light.max or 15) and tod < 18000 then
			changeHP(self, -1)
		elseif self.last_llvl and self.last_llvl < (def.stats.light.min or 0) then
			changeHP(self, -2)
		end
	end

	-- Random sounds
	if def.sounds and def.sounds.random and def.sounds.random[current_mode] then
		local rnd_sound = def.sounds.random[current_mode]
		if not self.snd_rnd_time then
			self.snd_rnd_time = rnd((rnd_sound.time_min or 5), (rnd_sound.time_max or 35))
		end
		if rnd_sound and self.soundtimer > self.snd_rnd_time + rnd() then
			self.soundtimer = 0
			self.snd_rnd_time = nil
			core.sound_play(rnd_sound.name, {
				pos = current_pos,
				gain = rnd_sound.gain or 1,
				max_hear_distance = rnd_sound.distance or 30,
				})
		end
	end

	self.mode = current_mode
end


cmer.get_staticdata = function(self)
	return {
		nametag = self.nametag,
		owner = self.owner,
		hp = self.object:get_hp(),
		mode = self.mode,
		tamed = self.tamed,
		modetimer = self.modetimer,
		lifetimer = self.lifetimer,
		soundtimer = self.soundtimer,
		fall_dist = self.fall_dist,
		in_water = self.in_water,
	}
end
